drivers/volume/portworx/portworx.go

```diff
@@ -2166,38 +2164,127 @@ func (d *portworx) CollectDiags(n node.Node) error {
 		if err != nil {
 			return fmt.Errorf("failed to collect diags on node %v, Err: %v %v", pxNode.Hostname, err, out)
 		}
+
 		logrus.Debugf("Successfully collected diags on node %v", pxNode.Hostname)
 		return nil
 	}
 
 	url := fmt.Sprintf("http://%s:9014", n.Addresses[0])
 
-	r := &DiagRequestConfig{
-		DockerHost:    "unix:///var/run/docker.sock",
-		OutputFile:    "/var/cores/diags.tar.gz",
-		ContainerName: "",
-		Profile:       false,
-		Live:          true,
-		Upload:        false,
-		All:           true,
-		Force:         true,
-		OnHost:        true,
-		Extra:         false,
-	}
-
 	c, err := client.NewClient(url, "", "")
 	if err != nil {
 		return err
 	}
-	req := c.Post().Resource(pxDiagPath).Body(r)
+	req := c.Post().Resource(pxDiagPath).Body(config)
+
 	resp := req.Do()
 	if resp.Error() != nil {
 		return fmt.Errorf("failed to collect diags on node %v, Err: %v", pxNode.Hostname, resp.Error())
 	}
+
+	if diagOps.Validate {
+		cmd := fmt.Sprintf("test -f %s", config.OutputFile)
+		out, err := d.nodeDriver.RunCommand(n, cmd, opts)
+		if err != nil {
+			return fmt.Errorf("failed to locate diags on node %v, Err: %v %v", pxNode.Hostname, err, out)
+		}
+
+		logrus.Debug("Validating CCM health")
+		// Change to config package.
+		url := fmt.Sprintf("http://%s:%d/1.0/status/troubleshoot-cloud-connection", n.MgmtIp, 1970)
+		resp, err := http.Get(url)
+		if err != nil {
+			return fmt.Errorf("failed to talk to CCM on node %v, Err: %v", pxNode.Hostname, err)
+		}
+		defer resp.Body.Close()
+
+		/*
+			// Check S3 bucket for diags
+
+			// TODO: Waiting for S3 credentials.
+		*/
+	}
+
 	logrus.Debugf("Successfully collected diags on node %v", pxNode.Hostname)
 	return nil
 }
 
+func (d *portworx) CollectAsyncDiags(n node.Node, config *torpedovolume.DiagRequestConfig) error {
+	diagsMgr := d.getDiagsManager()
+	jobMgr := d.getDiagsJobManager()
+
+	req := &api.SdkDiagsCollectRequest{
+		Issuer:      "CLI",
+		ProfileOnly: config.Profile,
+		Live:        config.Live,
+	}
+
+	req.Node = &api.DiagsNodeSelector{
+		NodeIds: []string{n.Id},
+	}
+
+	resp, err := diagsMgr.Collect(d.getContext(), req)
+	if err != nil {
+		return err
+	}
+	if resp.Job == nil {
+		err = fmt.Errorf("diags collection request submitted but did not get a Job ID in response")
+		return err
+	}
+
+	for {
+		resp, _ := jobMgr.GetStatus(d.getContext(), &api.SdkGetJobStatusRequest{
+			Id:   resp.Job.GetId(),
+			Type: resp.Job.GetType(),
+		})
+
+		state := resp.GetJob().GetState()
+
+		if state == api.Job_DONE || state == api.Job_FAILED || state == api.Job_CANCELLED {
+			break
+		}
+		fmt.Println("Waiting 5 seconds to check job status again.")
+		// Sleep 5 seconds until we check jobs again.
+		time.Sleep(5 * time.Second)
+	}
```

Comment:
From client standpoint, it needs to timeout at some high value. Let's time out at 15 minutes.